ARCs/arc-0060.md

```diff
@@ -0,0 +1,330 @@
+
+## Specification
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**MAY**", and "**OPTIONAL**" in this document are to be interpreted as described in <a href="https://www.ietf.org/rfc/rfc2119.txt">RFC-2119</a>.
+
+> Comments like this are non-normative
+
+### Overview
+
+> This section is non-normative
+
+Signatures of bytes are processed with the `signData(data, metadata, signer)` function.
+
+`data` is a string representing the structured data being signed. It is the representation of a JSON Schema, provided with the `metadata`.
+
+`metadata` is a `StdSignMetadata` object that describes the signature scope and the type of data being signed, including the encoding and the JSON schema.
+
+`signer` is a byte array representing the public key to use for the signing operation.
+
+### Interfaces
+
+> Interfaces are defined in TypeScript. All the objects that are defined are valid JSON objects.
+
+ARC-60 uses interchangeably the terms "throw an error" and "reject a promise with an error".
+
+#### **Interface `SignDataFunction`**
+
+A wallet function `signData` is defined by the interface:
+
+```tsx
+export type SignDataFunction = {
+    data: StdData,
+    metadata: StdSignMetadata,
+    signer: Ed25519Pk,
+}
+    => Promise<(SignedDataStr | null)>;
+```
+
+- `data` is a `StdData` object representing the structured data being signed (defined below).
+- `metadata` is a `StdSignMetadata` objects (defined below) that provide additional information on the data being signed, such as the `data` encoding and the JSON schema.
+- `signer` is a `Ed25519Pk` object (defined below) that represents the signer public key.
+
+The `signData` function returns a `SignedDataStr` object or, in case of error, rejects the promise with an error object `SignDataError`.
+
+#### Interface `StdData`
+
+The `StdData` object is a string that comply with the `metadata` JSON schema. It **MUST** be the byte representation of a canonicalized JSON object, following the <a href="https://www.rfc-editor.org/rfc/rfc8785">RFC 8785</a>.
+
+```tsx
+export type StdData = string;
+```
+
+The `StdData` must be validated with respect the JSON schema provided with `metadata` (defined below).
+
+#### Interface `Ed25519Pk`
+
+An `Ed25519Pk` object is a 32-byte public key, point of the ed25519 elliptic curve. The key **MUST NOT** be transformed into an Algorand address.
+
+```tsx
+export type Ed25519Pk = Uint8Array;
+```
+
+> The wallet **MAY** want to operate with standard Algorand address directly. The transformation from a generic key to an Algorand address is left to the implementation. See <a href="https://developer.algorand.org/docs/get-details/accounts/#transformation-public-key-to-algorand-address">Public Key to Algorand Address</a> section of the developer portal.
+
+#### Interface `SignedDataStr`
+
+`SignedDataStr` is the produced 64-byte array, ed25519 digital signature, of the signed data.
```

Comment:
According to this definition, `SigendDataStr` is purely the signature (64 bytes long). In the reference implementation, [`nacl.sign`](https://www.npmjs.com/package/tweetnacl#naclsignmessage-secretkey) is used which returns the signature _and_ the original message.

If `SignedDataStr` represents purely the signature, wouldn't something like `Signature` be a more appropriate name for this type?

Reply:
Good catch. Here https://github.com/algorandfoundation/ARCs/pull/284/commits/b244ed632ed393a0b33fea1fb0f6c4de03e74506 I updated both the naming and the reference implementation. I am using [`nacl.sign.detached`](https://www.npmjs.com/package/tweetnacl#naclsigndetachedmessage-secretkey) now.